brooklyn-server/core/src/main/java/org/apache/brooklyn/core/entity/AbstractApplication.java

```diff
@@ -220,13 +221,34 @@ public void stop() {
                 //but that could be handled by the impl at management
                 //(keeping recently unmanaged things)  
                 //  however unmanaging must be done last, _after_ we stop children and set attributes 
+                Location location = Iterables.getFirst(getLocations(), null);
                 getEntityManager().unmanage(this);
+                unmanageProvisionerIfNotUsed(location);
             }
         }
 
         logApplicationLifecycle("Stopped");
     }
 
+
+    private boolean unmanageProvisionerIfNotUsed(/*MachineProvisioningLocation<MachineLocation>*/Location provisioner) {
+        if (provisioner.getChildren().isEmpty()) {
+            Entity parentEntity = getParent();
+            if (parentEntity == null) {
+                getManagementContext().getLocationManager().unmanage(provisioner);
+                return true;
+            } else {
+                /*
```

Comment:
Else block code is commented out. What is the purpose of this section? Is it complete?